raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java

```diff
@@ -194,14 +196,50 @@ private void completeCurrentBatch() {
         MemoryRecords data = currentBatch.build();
         completed.add(new CompletedBatch<>(
             currentBatch.baseOffset(),
-            currentBatch.records(),
+            Optional.of(currentBatch.records()),
             data,
             memoryPool,
             currentBatch.initialBuffer()
         ));
         currentBatch = null;
     }
 
+    public void appendLeaderChangeMessage(LeaderChangeMessage leaderChangeMessage, long currentTimeMs) {
+        appendLock.lock();
+        try {
+            maybeCompleteDrain();
```

Comment:
I think I was a little off in my suggestion to add this. I don't think this is sufficient to ensure the current batch gets completed before we add the leader change message since `maybeCompleteDrain` will only do so if a drain has been started. Maybe the simple thing is to call `flush()`?